test/test_sum_product.py

```diff
@@ -1182,6 +1192,74 @@ def test_modified_partial_sum_product_16(sum_op, prod_op, vars1, vars2,
     assert_close(actual, expected, atol=5e-4, rtol=5e-4)
 
 
+@pytest.mark.parametrize('use_lazy', [True, False])
+@pytest.mark.parametrize('vars1,vars2', [
+    (frozenset(),
+     frozenset({"time", "x_prev", "x_curr", "y_curr", "z0", "z1", "z2"})),
+    (frozenset({"y_curr", "z0", "z1", "z2"}),
+     frozenset({"time", "x_prev", "x_curr"})),
+    (frozenset({"time", "x_prev", "x_curr", "y_curr", "z0", "z1", "z2"}),
+     frozenset()),
+])
+@pytest.mark.parametrize('x_dim,y_dim,z_dim,time', [
+    (2, 3, 2, 5), (1, 3, 2, 5), (2, 1, 2, 5), (2, 3, 2, 1),
+])
+@pytest.mark.parametrize('sum_op,prod_op', [(ops.logaddexp, ops.add), (ops.add, ops.mul)])
+def test_modified_partial_sum_product_17(use_lazy, sum_op, prod_op, vars1, vars2,
+                                         x_dim, y_dim, z_dim, time):
+
+    f1 = random_tensor(OrderedDict({}))
+
+    f2 = random_tensor(OrderedDict({
+        "time": Bint[time],
+        "x_prev": Bint[x_dim],
+        "x_curr": Bint[x_dim],
+    }))
+
+    f3_1 = random_tensor(OrderedDict({
+        "time": Bint[time],
+        "x_curr": Bint[x_dim],
+        "y_curr": Bint[y_dim],
+        "z0": Bint[z_dim],
+        "z1": Bint[z_dim],
+    }))
+
+    f3_2 = random_tensor(OrderedDict({
+        "time": Bint[time],
+        "x_curr": Bint[x_dim],
+        "y_curr": Bint[y_dim],
+        "z1": Bint[z_dim],
+        "z2": Bint[z_dim],
+    }))
+
+    f3_3 = random_tensor(OrderedDict({
+        "time": Bint[time],
+        "x_curr": Bint[x_dim],
+        "y_curr": Bint[y_dim],
+        "z2": Bint[z_dim],
+    }))
+
+    factors = [f1, f2, f3_1, f3_2, f3_3]
+    plate_to_step = dict({"time": {"x_prev": "x_curr"}})
+
+    with interpretation(lazy if use_lazy else eager):
+        factors1 = modified_partial_sum_product(sum_op, prod_op, factors, vars1, plate_to_step)
+        factors2 = modified_partial_sum_product(sum_op, prod_op, factors1, vars2, plate_to_step)
+        actual = reduce(prod_op, factors2)
+    actual = apply_optimizer(actual)
+
+    local_var_dict = {"time": frozenset({"y_curr", "z0", "z1", "z2"})}
+    local_markov_var_dict = {"time": frozenset()}
+    global_vars = frozenset()
+    markov_to_step = dict({"time": {"x"}})
```

Comment:
Should this be `{"time": {"x_prev": "x_curr"}}` or `{"time": {}}`? I wouldn't expect `Dict[str, Set[str]]` to be a supported datatype for this arg.

Reply:
I'm not sure, although note that this argument is only passed to the test helper `_expected_modified_partial_sum_product`.

Probably the right thing to do here is to implement a general `funsor.sum_product.unroll` function that takes the same arguments as `funsor.sum_product.{modified_}partial_sum_product` and returns a new set of arguments specifying an equivalent sum-product problem with no plates or Markov dimensions and use that to replace the custom unrolling logic in this test helper.

Reply:
`_expected_modified_partial_sum_product` assumes that markov vars have a `_prev` and `_curr` name convention and converts `x` from `{"time": {"x"}}` to a pair of `x_prev` and `x_curr`. It was simpler to quickly write unrolling logic for nested markov dims this way. I agree that there should be a general and  better documented `funsor.sum_product.unroll` function.

Reply:
Thanks for explaining, that makes sense.